File: EstruturaDeDados/ordenacoes/src/main/java/mergeSort/MergeSort.java
package mergeSort;

import funcoes.FuncoesBasicas;

public class MergeSort {
    public static void main(String args[]) {

        FuncoesBasicas funcoesBasicas = new FuncoesBasicas();

        int[] vetor = funcoesBasicas.seedVetor();

        funcoesBasicas.printArray(vetor, false);

        mergeSort(vetor, vetor.length);

        funcoesBasicas.printArray(vetor, true);

    }

    public static void mergeSort(int[] a, int n) {
        if (n < 2) {
            return;
        }

        int meio = n / 2;
        int[] esquerda = new int[meio];
        int[] direita = new int[n - meio];

        for (int i = 0; i < meio; i++) {
            esquerda[i] = a[i];
        }

        for (int i = meio; i < n; i++) {
            direita[i - meio] = a[i];
        }

        mergeSort(esquerda, meio);
        mergeSort(direita, n - meio);

        merge(a, esquerda, direita, meio, n - meio);
    }

    public static void merge(int[] a, int[] l, int[] r, int left, int right) {
        int i = 0, j = 0, k = 0;

        while (i < left && j < right) {
            if (l[i] <= r[j]) {
                a[k++] = l[i++];
            } else {
                a[k++] = r[j++];
            }
        }

        while (i < left) {
            a[k++] = l[i++];
        }

        while (j < right) {
            a[k++] = r[j++];
        }
    }

    static void printArray(int arr[]) {
        for (int i = 0; i < arr.length; ++i) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
}

File: EstruturaDeDados/ordenacoes/src/main/java/mergeSort/MergeSort.py
import random

def seedVetor():
    tamanho = random.randint(20, 50)
    vetor = [random.randint(1, 100) for _ in range(tamanho)]
    return vetor

def printArray(arr, sorted):
    print()
    print("ordenado?", sorted)
    for num in arr:
        print(num)

def mergeSort(a):
    if len(a) < 2:
        return a

    meio = len(a) // 2

    left = mergeSort(a[:meio])
    right = mergeSort(a[meio:])

    return merge(left, right)

def merge(left, right):
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])

    return result

vetor = seedVetor()
printArray(vetor, False)

vetor = mergeSort(vetor)

printArray(vetor, True)



